File: java/src/main/java/yatzy/Yatzy.java
package yatzy;

import java.util.Arrays;
import java.util.Comparator;
import java.util.HashMap;
import java.util.Map;
import java.util.Optional;

public class Yatzy {

	public static int chance(int[] values) {
		return Arrays.stream(values).sum();
	}

	public static int yatzy(int[] values) {
		int score = 0;
		Map<Integer, Integer> occurences = getOccurencesByValue(values);
		if (occurences.size() == 1) {
			score = 50;
		}
		return score;
	}

	public static int ones(int[] values) {
		return sumAllValuesOf(1, values);
	}

	public static int twos(int[] values) {
		return sumAllValuesOf(2, values);
	}

	public static int threes(int[] values) {
		return sumAllValuesOf(3, values);
	}

	public static int fours(int[] values) {
		return sumAllValuesOf(4, values);
	}

	public static int fives(int[] values) {
		return sumAllValuesOf(5, values);
	}

	public static int sixes(int[] values) {
		return sumAllValuesOf(6, values);
	}

	public static int score_pair(int[] values) {
		Map<Integer, Integer> occurences = getOccurencesByValue(values);
		Optional<Integer> score = occurences.entrySet().stream().filter(e -> e.getValue().intValue() == 2)
				.map(t -> t.getKey()).max(Comparator.naturalOrder());

		return score.isPresent() ? score.get().intValue() * 2 : 0;

	}

	public static int two_pair(int[] values) {
		int score = 0;
		Map<Integer, Integer> occurences = getOccurencesByValue(values);
		long occurenceOfTwo = occurences.entrySet().stream().filter(e -> e.getValue().intValue() >= 2).count();
		if (occurenceOfTwo == 2) {
			score = occurences.entrySet().stream().filter(e -> e.getValue().intValue() >= 2).map(t -> t.getKey())
					.reduce((a, b) -> a + b).get();
		}
		return score * 2;
	}

	public static int four_of_a_kind(int[] values) {
		int score = 0;
		Map<Integer, Integer> occurences = getOccurencesByValue(values);

		long numberOfDiceWith4TimesSameNumber = occurences.entrySet().stream().filter(e -> e.getValue().intValue() >= 4)
				.count();

		if (numberOfDiceWith4TimesSameNumber >= 1) {
			score = occurences.entrySet().stream().filter(e -> e.getValue().intValue() >= 4).mapToInt(t -> t.getKey())
					.findFirst().getAsInt() * 4;

		}

		return score;
	}

	public static int three_of_a_kind(int[] values) {
		int score = 0;
		Map<Integer, Integer> occurences = getOccurencesByValue(values);
		long numberOfDiceWith3AtLeastTimesSameNumber = occurences.entrySet().stream()
				.filter(e -> e.getValue().intValue() >= 3).count();

		if (numberOfDiceWith3AtLeastTimesSameNumber >= 1) {
			score = occurences.entrySet().stream().filter(e -> e.getValue().intValue() >= 3).mapToInt(t -> t.getKey())
					.findFirst().getAsInt() * 3;
		}

		return score;
	}

	public static int smallStraight(int[] values) {
		boolean hasAllTheDistinctValues = Arrays.stream(values).distinct().count() == values.length;
		int score = 0;
		if (hasAllTheDistinctValues) {
			score = Arrays.stream(values).distinct().sum();
		}
		return score;

	}

	public static int largeStraight(int[] values) {
		int score = 0;
		int[] largeStraight = { 2, 3, 4, 5, 6 };
		Arrays.sort(largeStraight);
		Arrays.sort(values);
		if (Arrays.equals(values, largeStraight)) {
			score = Arrays.stream(values).sum();
		}
		return score;
	}

	public static int fullHouse(int[] values) {
		int score = 0;
		Map<Integer, Integer> occurences = getOccurencesByValue(values);
		boolean hasTwoOfKind = occurences.entrySet().stream().filter(e -> e.getValue().intValue() == 2).count() > 0;
		boolean hasThreeOfKind = occurences.entrySet().stream().filter(e -> e.getValue().intValue() == 3).count() > 0;
		if (hasTwoOfKind && hasThreeOfKind) {
			score = occurences.entrySet().stream().mapToInt(k -> k.getKey() * k.getValue()).sum();
		}
		return score;

	}

	private static Map<Integer, Integer> getOccurencesByValue(int[] values) {
		Map<Integer, Integer> occurences = new HashMap<Integer, Integer>();
		for (int i = 0; i < values.length; i++) {
			if (occurences.get(values[i]) == null) {
				occurences.put(values[i], 1);
			} else {
				occurences.put(values[i], occurences.get(values[i]).intValue() + 1);
			}

		}
		return occurences;
	}

	private static int sumAllValuesOf(Integer valueToSum, int[] values) {
		return Arrays.stream(values).filter(t -> t == valueToSum).sum();

	}
}

File: javascript/lib/yatzy.js
var Yatzy = function () {

    let _sumAllValues = function (dices) {
		return _sumAllValuesOf(dices);
	};

	let _sumAllValuesOf = function(dices,valueToSum) {
        var total = 0;
        for(let i = 0; i<dices.length;i++){
            if (!valueToSum || valueToSum == dices[i]) {
				total += dices[i];
			}
        }
        return total;
    };
    
    let _distinctValues = function (dices) {
		let distinctDices = [];
        for(let i = 0; i<dices.length;i++){
			if (distinctDices.indexOf(dices[i])==-1) {
				distinctDices.push(dices[i]);
			}
        }
        
		return distinctDices;
	}

	let _sumFirstValueOfKind = function (nbOfKind, dices) {
        let distinctDicesValues = _distinctValues(dices);

        for(let i = 0; i<distinctDicesValues.length;i++){
            let sum = _sumNbOccurenceOfValue(distinctDicesValues[i], nbOfKind, dices);
            if (sum > 0) {
                return sum;
            }
        }      
		return 0;
	}

	let _sumNbOccurenceOfValue = function (value,nbOccurence, dices) {
		if (((_sumAllValuesOf(dices,value)) / value) >= nbOccurence) {
			return value * nbOccurence;
		}
		return 0;
    }
    
    let _straight = function(dices) {
		return _distinctValues(dices).length == 5 ? _sumAllValues(dices) : 0;
	}

    return {

        ones: function () {
            return _sumAllValuesOf(arguments,1);
        }
        ,
        twos: function () {
            return _sumAllValuesOf(arguments,2);
      
        },

        threes: function () {
            return _sumAllValuesOf(arguments,3);
      
        },

        fours: function () {   
            return _sumAllValuesOf(arguments,4);
        }

        ,

        fives: function () {
            return _sumAllValuesOf(arguments,5);
        },

        sixes: function () {   
            return _sumAllValuesOf(arguments,6);
        },

        chance: function () {
            return _sumAllValues(arguments) ;
        },
        yatzy: function () {
            return _distinctValues(arguments).length == 1 ? 50 : 0;
        },

    
        score_pair: function () {
            var maxSumPair = 0;
            let distinctDicesValues = _distinctValues(arguments);
        
            
            for(let i = 0; i<distinctDicesValues.length;i++){
                let sumPairScore = _sumNbOccurenceOfValue(distinctDicesValues[i], 2, arguments);
                if (sumPairScore > maxSumPair) {
                    maxSumPair = sumPairScore;
                }
            }

            return maxSumPair;
        },

        two_pair : function () {
            var total = 0;
            let distinctDicesValues = _distinctValues(arguments);

            for(let i = 0; i<distinctDicesValues.length;i++){
                total += _sumNbOccurenceOfValue(distinctDicesValues[i], 2, arguments);
            }
            return total;
        },

        
        three_of_a_kind: function () {
            return _sumFirstValueOfKind(3, arguments);
        },

        four_of_a_kind: function () {
            return _sumFirstValueOfKind(4, arguments);
        },

        smallStraight: function () {
           return _straight(arguments);
        },

        largeStraight: function (dices) {
            return _straight(arguments);
        },

        fullHouse: function () {
            return _distinctValues(arguments).length == 2 ? _sumAllValues(arguments) : 0;
        }
    }
}();

module.exports = Yatzy;
